src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java

```diff
@@ -0,0 +1,118 @@
+ * similarity of two samples. It was independently developed by the botanists
+ * Thorvald Sørensen and Lee Raymond Dice, who published in 1948 and 1945
+ * respectively. The index is known by several other names, especially
+ * Sørensen–Dice index, Sørensen index and Dice's coefficient. Other variations
+ * include the "similarity coefficient" or "index", such as Dice similarity
+ * coefficient (DSC).
+ * </p>
+ *
+ * <p>
+ * This implementation is based on the Sørensen–Dice similarity algorithm from
+ * <a href=
+ * "http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient">
+ * http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient</a>.
+ *
+ *
+ * </p>
+ *
+ * @since 1.7
+ */
+public class SorensenDicesSimilarity implements SimilarityScore<Double> {
+
+    /**
+     *
+     * <pre>
+     * similarity.apply("", "")                     = 1.0
+     * similarity.apply("foo", "foo")               = 1.0
+     * similarity.apply("foo", "foo ")              = 0.8
+     * similarity.apply("foo", "foo ")              = 0.66
+     * similarity.apply("foo", " foo ")             = 0.66
+     * similarity.apply("foo", " foo")              = 0.66
+     * similarity.apply("", "a")                    = 0.0
+     * similarity.apply("aaapppp", "")              = 0.0
+     * similarity.apply("frog", "fog")              = 0.4
+     * similarity.apply("fly", "ant")               = 0.0
+     * similarity.apply("elephant", "hippo")        = 0.0
+     * similarity.apply("hippo", "elephant")        = 0.0
+     * similarity.apply("hippo", "zzzzzzzz")        = 0.0
+     * similarity.apply("hello", "hallo")           = 0.5
+     * similarity.apply("ABC Corporation", "ABC Corp") = 0.7
+     * similarity.apply("D N H Enterprises Inc", "D &amp; H Enterprises, Inc.") = 0.74
+     * similarity.apply("My Gym Children's Fitness Center", "My Gym. Childrens Fitness") = 0.81
+     * similarity.apply("PENNSYLVANIA", "PENNCISYLVNIA") = 0.69
+     * </pre>
+     *
+     * @param left  the first CharSequence, must not be null
+     * @param right the second CharSequence, must not be null
+     * @return result similarity
+     * @throws IllegalArgumentException if either CharSequence input is {@code null}
+     */
+    @Override
+    public Double apply(final CharSequence left, final CharSequence right) {
+
+        if (left == null || right == null) {
+            throw new IllegalArgumentException("CharSequences must not be null");
+        }
+
+        if (left.equals(right)) {
+            return 1d;
+        }
+
+        if (left.length() == 0 || right.length() == 0) {
+            return 0d;
+        }
+
+        Set<String> nLeft = createBigrams(left);
+        Set<String> nRight = createBigrams(right);
+
+        final int total = nLeft.size() + nRight.size();
+        nLeft.retainAll(nRight);
+        final int intersection = nLeft.size();
+
+        return (2.0d * intersection) / total;
```

Comment:
Hummm, I think we have a bug here. In line ~88, we are checking if the length of both are zero. If so, we return a coefficient = 0d.

But then here `total` is the sum of the size of the left and right sets of bigrams. But for an input such as `left="a"`, and `right="b"`, we have zero bigrams.

Zero bigrams, gives a `total=0`. And then when we divide by 0, we get a `NaN`. Probably we need to at least document it... but I would prefer to investigate whether we have a value in the algorithm to be returned (i.e. when no bigrams found, what's in the paper/Wikipedia page/other implementations? What makes sense to users?).

This also needs to be covered by a test (even though we have 100% line coverage, it doesn't hurt to cover the specific corner cases in our API :grin: ).

Reply:
This is my biggest nit with the PR. The Sorensen-Dice similarity is readily computed from the Jaccard. As I said in my first comment:

`The score is related to the Jaccard score and can be computed directly from it (S = 2J / (1 + J)).`

So the novel part of this class is the use of bigrams. This could easily be applied to the `JaccardSimilary` class already in the library.

So perhaps change to a new class `NGramJaccardSimilarity` that accepts a size argument for the constructor for the size of the N-gram. Then just compute the Jaccard. This should: 

- throw when size is 0
- compute the same score as the `JaccardSimilarity` class when the size is 1
- compute the bigram score when the size is 2, etc
- return 0 when any input `CharSequence` is shorter than the bigram length

Reply:
I read in the Wikipedia article I believe, that it was possible to calculate it from the Jaccard value. But hadn't thought about any reusing any code, which might be a good idea.

>So perhaps change to a new class NGramJaccardSimilarity that accepts a size argument for the constructor for the size of the N-gram. Then just compute the Jaccard.

Would it work, to keep both JaccardSimilarity, and this new SorensenDiceSimilarity. But re-use the Jaccard within the Sorensen?

This way we wouldn't need to check the size of the N-gram. Though I am not looking at the code right now, so can't confirm if that wouldn't make the code harder to understand.